clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -515,20 +516,32 @@ private static int parseAcks(String acksString) {
     /**
      * Wait for cluster metadata including partitions for the given topic to be available.
      * @param topic The topic we want metadata for
+     * @param partition A specific partition expected to exist in metadata, or null if there's no preference
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
      * @return The cluster containing topic metadata and the amount of time we waited in ms
      */
-    private ClusterAndWaitTime waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long maxWaitMs) throws InterruptedException {
+        if (partition != null && partition < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid partition given with record: %d. Partition number should always greater than 0.", partition));
+        }
+
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
         Cluster cluster = metadata.fetch();
-        if (cluster.partitionsForTopic(topic) != null)
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        // Return cached metadata if we have it, and if the record's partition is either undefined
+        // or within the known partition range
+        if (partitions != null && (partition == null || partition < partitions.size()))
             return new ClusterAndWaitTime(cluster, 0);
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
-        long elapsed = 0;
-        while (cluster.partitionsForTopic(topic) == null) {
+        long elapsed;
+        // If no metadata was available, issue metadata update requests until we get successful response or we timeout.
+        // If a cached copy of metadata was present but the requested partition was out of bounds, issue a single
+        // request to try to refresh metadata, in case a partition expansion has taken place.
```

Comment:
Just keep in mind that it's important to be careful not to simplify to the point of being incorrect (e.g. a successful response is not the same as a response containing metadata for a given topic).